### ISSUE TITLE
Thermostat doesn't turn on on set-temperature

### ISSUE THREAD

**Issue by 3v1n0** [nielsfaber/scheduler-component#142]
### Checklist
- [x] I checked for similar existing issues (both open and closed) before posting.
- [x] I am running the latest version.

### Expected behavior
[What you expected to happen]

When scheduling to set temperature to a given value to my thermostat, it sets the temperature and turns it on if needed.

### Actual behavior
[What actually happened]

It sets the temperature, but doesn't turn on the heater.

### Steps to Reproduce
[Explain how to see the behaviour]

My heater is a tuya one that works with new tuya v2 platform (it used to work properly in this way), and its attributes are:

```yaml
hvac_modes:
  - 'off'
  - heat_cool
min_temp: 5
max_temp: 35
target_temp_step: 0.5
current_temperature: 18
temperature: 19
supported_features: 1
```

I've configured it in this way: ![image](https://user-images.githubusercontent.com/345675/140513191-d6ee96c3-0c1b-4183-acea-302e9e06beb1.png)

So that it turns off and after some time it turns on, but as you can see it didn't turn on, while the temperature is properly set.
And I don't have an UI button to turn it on (but it should be implicit in not being off).

**Comment by mdeweerd** [comment]
I am not opening a new issue, mine seems to be related.

# Summary
A schedule is supposed to act at 4AM to enable enable heating and set the temperature to 8°C.  Only the heating is set to the correct temperature.  

![image](https://user-images.githubusercontent.com/1504752/141281043-d5409ffb-3d11-452e-91bc-e09b587514ec.png)

# Detail
I have three Generic Thermostats in my system - Zigbee Temperature Sensors are used to control on/of relays.

Every day at 4AM, I programmed the scheduler to set the temperature to 8°C and to activate the Generic Thermostats.  The goal is to ensure that the temperature is set to a minimum and that the heating is on regardless of what happened during the day.

The schedule does set the Thermostat in heating mode, but does not set the temperature as the relays are actually enabled to increase the tempearture to the setpoints set at the end ov the day:
(top part= heaters active when green, bootom = tempeature measured (all above 14.5°C - the times are not perfectly aligned due to a graph "bug", but heating starts at 4AM):
![image](https://user-images.githubusercontent.com/1504752/141281510-aaa96fea-4e57-4f5f-8b2b-89768828a88d.png)

The schedule is called "Hors gel".

This is an extract from the journal:
![image](https://user-images.githubusercontent.com/1504752/141282269-9b809326-f4e1-42c9-843a-59bee77a2abf.png)

I am under the impression that only climate.set_hvac_mode is called and not set temperature.  Idealy climate.set_temperature is called just before setting the hvac_mode when switching on, and when switching off, set it just after changing the hvac mode (to avoid unnecessary heating if there is some delay between the settings).

**Comment by mdeweerd** [comment]
climate.set_temperature can also set the HVAC mode, so then there is no issue in the order the mode and temperature are set ;-).
I suggest to switch to climate.set_temperature.

**Comment by nielsfaber** [comment]
If you use the ‘heat’ action, scheduler will set the hvac_mode to ‘heat’ and 1 second later it sets the temperature setpoint to the value selected in the card.
This works fine for most users.

It could be that this doesn’t work with your climate device, either because it should not be in the ‘heat’ mode (some devices are 100% continuously ON at this mode) or because the 1 second delay is too short (there is also another open issue about this).


Please find out what is the cause. You could create a script or automation which performs the same sequence, and make alterations from there.

**Comment by mdeweerd** [comment]
This is the generic thermostat implementation of the Home Assistant core, so the temperature setting is managed on HA itself, not sent to the devices.  HA just receives the temperature from the device and then sends on/off commands to the relay.

Here is an extract from last night's journal.
We can see that "Hors gel" is triggered at 4:00AM and at 4:01AM  which corresponds to the initial trigger and the 1 minute delay.
I suppose that the services "set_hvac_mode" are called by the scheduler, and probably also the service events "climate.set_temperature" at 4:05 and 4:07 (and 04:14:06 not shown here).  

![image](https://user-images.githubusercontent.com/1504752/141683792-5ab98bea-6174-4c8c-b187-a08013492d37.png)

As indicated, the delay is strange, my Generic thermostat definitions look like this:
```
climate:
  - platform: generic_thermostat
    name: Principal
    heater: light.tz3000_5ng23zjs_ts0001_8d421efe_on_off
    target_sensor: sensor.principal_temperature
    min_temp: 7
    max_temp: 23
    ac_mode: false
    target_temp: 20.5
    min_cycle_duration: 60
    cold_tolerance: 0.3
    hot_tolerance: 0.2
    initial_hvac_mode: 'off'
    # away_temp: 7
    precision: 0.1
```

I think only the "climate.set_temperature" service is needed as it also allows to set the hvac mode (at least in the developer UI).

As this is not an automation as defined by HA, we can't see the graphs of what triggered and what was actioned, I do not know what is logged (I could activate logging).

I can have a look at the code, but not in the very near future.

**Comment by nielsfaber** [comment]
@mdeweerd Thanks for your input!
Generic thermostat should work without problems, in fact I use it for testing scheduler myself as well.

Indeed it is possible to set the hvac_mode in the climate.set_temperature service, but scheduler does NOT use this since 50%+ of all climate integrations will ignore this parameter.
The generic_thermostat integration is no exception, see code here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/generic_thermostat/climate.py#L364

This is the main reason why scheduler calls the set_temperature and set_hvac_mode separately.
Note that this is only true when the ‘heat’ action is configured for the schedule (not ‘set temperature’).
Could you share your schedule configuration (from developer tools -> states)?

Just thinking out loud: perhaps the problem occurs only when multiple climate devices are configured? Frankly I don’t know exactly how the 1 second delay behaves with 3 entities. Could you try the same for 1 device?
I would test this myself but I’m not at home at the moment 😅

**Comment by mdeweerd** [comment]
HA is great (when technical people can set it up), but the code you refer to in `climate.py` is disappointing.   It also looks like the minimum and maximum set point values are not verified when setting the temperature.

I understand that you do both the hvac_mode setting and set_temperature setting, but you could still send temperature and mode throught the set_temperature service, and still send the hvac_mode setting seperately.

I'll do some more tests - maybe you could add information about how to activate the logging to your README.md - that would help to enable it.

**Comment by nielsfaber** [comment]
So I tried with this schedule:
![image](https://user-images.githubusercontent.com/16861605/141736537-83a00e2a-19ce-43c4-aa03-66be94898e3b.png)

I set it to 'off' before 8:00 to make the correct test case.
At 8:00 the debug logs say:
```
2021-11-15 08:00:00 DEBUG (MainThread) [custom_components.scheduler.switch] Schedule c25006 is triggered, proceed with actions,
2021-11-15 08:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [c25006]: Executing service climate.set_hvac_mode on entity climate.generic_thermostat,
2021-11-15 08:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [c25006]: Executing service wait,
2021-11-15 08:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [c25006]: state of climate.generic_thermostat has changed, re-evaluating actions,
2021-11-15 08:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [c25006]: Executing service climate.set_temperature on entity climate.generic_thermostat
```

Which is as I would expect it to be.

The history shows:
![image](https://user-images.githubusercontent.com/16861605/141736694-8c087220-1f32-4f71-8a76-0638f95a19c0.png)

And the logbook shows:
![image](https://user-images.githubusercontent.com/16861605/141736782-87b9d77d-e679-4dcc-b415-991da9d12588.png)


So all works fine for me with the `generic_thermostat` integration as my test candidate.
I will repeat the test with a second entity, to verify this also works as intended.

**Comment by nielsfaber** [comment]
On second thought I see that the 1 second delay is not maintained since `state of climate.generic_thermostat has changed, re-evaluating actions` causes the execution loop to be interrupted.
I will make some modifications to handle this correctly, hopefully it improves things for you.

**Comment by mdeweerd** [comment]
Excellent - I have enabled debug logging in this HA setup without changing the configuration (and I upgraded to the latest published version of the scheduler).

So that way we'll get the logs for how it works now, and then I can try applying the schedule to only one device.

But it seems that you identified the cause ;-).

**Comment by mdeweerd** [comment]
Hi

I'm back with the log from last night - it suggests that all actions are done in the same second.  The schedule itself is rescheduled one minute after it was triggered.
```
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.switch] Schedule 46f187 is triggered, proceed with actions
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_hvac_mode on entity climate.reunion
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service wait
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: state of climate.reunion has changed, re-evaluating actions
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_temperature on entity climate.reunion
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_hvac_mode on entity climate.principal
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service wait
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: state of climate.reunion has changed, re-evaluating actions
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_temperature on entity climate.principal
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_hvac_mode on entity climate.bureau
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service wait
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: state of climate.principal has changed, re-evaluating actions
-16 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_temperature on entity climate.bureau
-16 04:01:00 DEBUG (MainThread) [custom_components.scheduler.timer] Timer of 46f187 set for 2021-11-17 04:00:00+01:00
```

So according to this, all is triggered in the same second, but the logbook only shows the set_hvac_service calls at 4AM and the set_temperature much later as before 04:06:23 , 04:07:54, 04:14:30 .

I have now enabled debugging on "climate".

**Comment by mdeweerd** [comment]
I had a look at the scheduler code and as you said, the set_temperature the action sequence is supposed to set the temperature 1 second after the hvac setting.
However it is clearly set at a variable later time - the three set_temperature settings are very likely the result of the schedule at 4AM.

I have added debug logging for homeassistant.helpser.events.

I would reverse the order of the actions anyway and add both HVAC and TEMPERATURE to the set temperature service call for services that support it (or awaiting a correction for the generic thermostat).

I added debug messages to actions.py to know when the timer expired when the asynchronous action handler is entered, when the asynchronous action is started and when it completed.

**Comment by nielsfaber** [comment]
@mdeweerd I have made some changes to the code that handles the execution of the actions.
The changes allow the execution of sequences of actions for multiple entities, they are now handled independently.

The changes resolved the problems for me.
The log now shows:
```
2021-11-16 21:41:00 DEBUG (MainThread) [custom_components.scheduler.timer] Timer b58d74 has reached slot 0,
2021-11-16 21:41:00 DEBUG (MainThread) [custom_components.scheduler.switch] Schedule b58d74 is triggered, proceed with actions,
2021-11-16 21:41:00 DEBUG (MainThread) [custom_components.scheduler.actions] [b58d74]: Executing service climate.set_hvac_mode on entity climate.generic_thermostat,
2021-11-16 21:41:00 DEBUG (MainThread) [custom_components.scheduler.actions] [b58d74]: Postponing next action for 5 seconds,
2021-11-16 21:41:00 DEBUG (MainThread) [custom_components.scheduler.actions] [b58d74]: Executing service climate.set_hvac_mode on entity climate.generic_thermostat_2,
2021-11-16 21:41:05 DEBUG (MainThread) [custom_components.scheduler.actions] [b58d74]: Executing service climate.set_temperature on entity climate.generic_thermostat,
2021-11-16 21:41:05 DEBUG (MainThread) [custom_components.scheduler.actions] [b58d74]: Executing service climate.set_temperature on entity climate.generic_thermostat_2,
2021-11-16 21:41:05 DEBUG (MainThread) [custom_components.scheduler.actions] [b58d74]: finished execution of actions
```

Since you seem to be comfortable playing with the python code, I would like to ask you to locally update the `scheduler/actions.py` file to my latest changes: https://github.com/nielsfaber/scheduler-component/blob/main/custom_components/scheduler/actions.py

Ofcourse I am interested whether the update also resolves the issues for you 😉

**Comment by mdeweerd** [comment]
I know a whole bunch of languages - my fingers and toes do not suffice to count them, python is not my preferred language but I've been confronted with it since 2014 with increasing intensity.  I even generated VHDL from python a long time ago to see how that performed - somebody wrote a "backend" for it.

Anyway, I've replaced my action.py with the new one.

Apparently you've reproduced it, so that's a good thing.

**Comment by mdeweerd** [comment]
Congrats, the delay issue is fixed - the heaters heat for 5 seconds, not several minutes.
```
2021-11-17 04:00:00 DEBUG (MainThread) [custom_components.scheduler.timer] Timer 46f187 has reached slot 0
2021-11-17 04:00:00 DEBUG (MainThread) [custom_components.scheduler.switch] Schedule 46f187 is triggered, proceed with actions
2021-11-17 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_hvac_mode on entity climate.reunion
2021-11-17 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Postponing next action for 5 seconds
2021-11-17 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_hvac_mode on entity climate.principal
2021-11-17 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Postponing next action for 5 seconds
2021-11-17 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_hvac_mode on entity climate.bureau
2021-11-17 04:00:00 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Postponing next action for 5 seconds
2021-11-17 04:00:05 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_temperature on entity climate.reunion
2021-11-17 04:00:05 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_temperature on entity climate.principal
2021-11-17 04:00:05 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: Executing service climate.set_temperature on entity climate.bureau
2021-11-17 04:00:05 DEBUG (MainThread) [custom_components.scheduler.actions] [46f187]: finished execution of actions
2021-11-17 04:01:00 DEBUG (MainThread) [custom_components.scheduler.timer] Timer of 46f187 set for 2021-11-18 04:00:00+01:00
````

Could you update the code to do "set_temperature" first with both the temperature AND the HVAC setting.

The Zigbee integration does use those settings:
```
    async def async_set_temperature(self, **kwargs):
        """Set new target temperature."""
        low_temp = kwargs.get(ATTR_TARGET_TEMP_LOW)
        high_temp = kwargs.get(ATTR_TARGET_TEMP_HIGH)
        temp = kwargs.get(ATTR_TEMPERATURE)
        hvac_mode = kwargs.get(ATTR_HVAC_MODE)
```

And setting the temperature first would avoid the unnecessary heating (for 5 seconds now) and relay switching.
Then (and maybe only 1 second later as before), you could the HVAC configuration.

Idealy the order would depend on the actual commands: when switching HVAC off, you would send that first, but I think in the general case HVAC is switched on and the few seconds delay does not matter for the user, but it matters for the heater.

Thanks for fixing this, and for a non professional coder you've done a better than professional average job for the component and also for fixing this ;-).

**Comment by nielsfaber** [comment]
@mdeweerd Thanks for the quick feedback!

> Could you update the code to do "set_temperature" first with both the temperature AND the HVAC setting.

I could (and I heard you the first time as well 🙂 ).

The concerns/disadvantages I see for switching the order is that I saw in the past (i.e. received user feedback) are:

1. Some climate integrations revert to some 'default' temperature setpoint when switching `HVAC_MODE` from `off` -> `heat`
2. Some climate integrations ignore the `set_temperature` command completely whilst in `HVAC_MODE` `off`

This is a bit conflicting with HA climate definition so you could argue the issue is caused by the integration itself, and not responsibility for scheduler to take into account.

There is always this trade-off I have to make when making changes, but my feeling is that your case is much more common so I think I will make your proposed change (and let's see what feedback I get..).

> Then (and maybe only 1 second later as before), you could the HVAC configuration.

The delay used to be 1 second but I increased it to 5 due to user feedback. Apparently some integrations ignore a command while still busy processing previous command (see for example https://github.com/nielsfaber/scheduler-component/issues/136).

> Idealy the order would depend on the actual commands: when switching HVAC off, you would send that first, but I think in the general case HVAC is switched on and the few seconds delay does not matter for the user, but it matters for the heater.

Switching `HVAC_MODE` to `off` goes without delay and without setting a setpoint. The sequence only applies to `heat`, `cool` and `regulate` commands (as they are called in scheduler-card).

> Thanks for fixing this, and for a non professional coder you've done a better than professional average job for the component and also for fixing this ;-).

Thanks for the compliment! I do have quite the opposite feeling, I am far from proud of the code. 
I would definitely not compare it to a professional job due to a lack of review, regression testing, documentation, etc.
But since I wrote it myself I can make changes easily and hopefully keep users happy 👍

**Comment by mdeweerd** [comment]
Professionnaly code should be revised, regression tested, documented, etc.  But that more often not the case.  Your code could be improved with regards to that of course, but there is structure in it - not one big lump of code in one file, with meaningfull names, you know where to look for this kind of issue, and the UI is a nice approach.

I've done work on ICs that you have in your phone where all of the things you mentioned are done, we even had automated bug introduction in the hardware language desciptions to see if the bug was detected or not by the regressions.  
Regressions were "constrained random".  The languages used allow to write test cases by indicating that a is [0...255] and b is [0..255] but an even number, and it that the result has to be "a * b +50".  In practice it's more complex as we'ld verify USB with such methods (which is a complex protocol), etc.
The tests would then be run with a great number of variations on a and b, the best ones (for coverage) would be selected to be run on every regression.  
Bug insertion is something like replacing "a==b" with "a!=b", False or True, one bug at a time.  This detects that parts of the design were not properly verified or dead code (parts of the design not used ever, or not used in the actual product).  
I have never seen that done on SW.

Verification took roughly twice the effort of designing it, but when you are putting tens of millions of the same chip on the market you can't afford it to fail, and you can't just fix it through a SW upgrade (if you're lucky a SW upgrade can bypass the issue).

At the other end of the scale, it's barely tested, put it online ASAP and if it doesn't work, fix it ASAP.

You could say that its not professionnal, and I would say that it's not what the profession "mandates", but its what the one giving the order is willing to pay for and the risk he's (implicitally) accepting.  Paying 3 times the engineering cost to do it perfectly well, or paying only slightly more than 1/3rd and coping with the impact can still be less expensive overall.  
The more products are sold, the more users there are, the higher the risk and the higher the effort put in making it work.  Also the more people of different skillsets are working on a project, the more you need to automate the qualtity ensurance.

Anyway, that's quite some effort, and generally team work.

**Comment by nielsfaber** [comment]
@mdeweerd Thanks for your positive view on my work. Mostly I am confronted with bug reports or requests for new stuff, so it's nice to receive some good feedback 👍 

@3v1n0 I released a new version v3.2.3 with the changes discussed here. Could you please update and confirm that the issue is solved for you as well? You can close the issue if so.

**Comment by mdeweerd** [comment]
Updated, all is ok:
- set_temperature first;
- set_hvac_mode next.

I did not have any unnecessary heating this night ;-).
Expected consumption for today: 0.53kWh, consumption the other day: 1.33kWh.

I know it's not that much, but "Vele kleintjes maken een groot".

Bedankt!

I am not closing this yet, @3v1n0 did not provide any feedback, but the improvements can fix it - so let's close it in a few days if there's no further feedback.
